### PR TITLE
Handle private types for array indices

### PR DESCRIPTION
**Pull request by xbauch**
Includes clean-up of the array handling in general.

### CODE REVIEW

File: gnat2goto/driver/arrays.adb [hunk 1]
```diff
@@ -997,4 +997,41 @@ package body Arrays is
                                  Get_Name (Data_Member));
    end Get_Data_Member;
 
+   function Build_Array_Size (Array_Comp : Irep; Idx_Type : Irep) return Irep
+   is (Build_Array_Size (First      => Get_First_Index (Array_Comp),
+                         Last       => Get_Last_Index  (Array_Comp),
+                         Idx_Type   => Idx_Type));
```
Comment (hannes-steffenhagen-diffblue) [inline gnat2goto/driver/arrays.adb]: When you're both moving and changing large blocks of code it does usually make sense to do both in separate commits so it's easier to see what you've changed; In this case it's just between two files so it's not too bad

File: gnat2goto/driver/arrays.ads [hunk 2]
```diff
@@ -81,6 +92,12 @@ private
    function Make_Array_Index_Op
      (Base_Irep : Irep; Base_Type : Node_Id; Idx_Irep : Irep) return Irep;
 
+   function Build_Array_Size (First : Irep; Last : Irep; Idx_Type : Irep)
+                              return Irep
+     with Pre => (Kind (First) in Class_Expr
+                  and then Kind (Last) in Class_Expr
```
Comment (hannes-steffenhagen-diffblue) [inline gnat2goto/driver/arrays.ads]: FWIW my preference is to use `and` instead of `and then` if there's no order dependence of the expressions

File: gnat2goto/driver/arrays.adb [hunk 3]
```diff
@@ -762,36 +762,17 @@ package body Arrays is
                      Overflow_Check  => False,
                      I_Type          => Get_Type (Idx_Irep),
                      Range_Check     => False);
-      Result_Type : Irep;
-      Pointer_Type : constant Irep := New_Irep (I_Pointer_Type);
+
```
Comment (hannes-steffenhagen-diffblue) [inline gnat2goto/driver/arrays.adb]: Not part of this PR, but I'm wondering why we even need the `Base_Type` here? Shouldn't it be sufficient to have an `Irep` type?

File: gnat2goto/driver/arrays.adb [hunk 4]
```diff
@@ -87,13 +89,7 @@ package body Arrays is
          --  2: data pointer
          --  Using the component numbers may be dropped in the future or it
          --  may be enforced.
```
Comment (hannes-steffenhagen-diffblue) [inline gnat2goto/driver/arrays.adb]: Comment seems out of date now

File: testsuite/gnat2goto/tests/private_type_dec_array/test.out [hunk 5]
```diff
@@ -0,0 +1,2 @@
+[1] file private_dec_array.adb line 6 assertion: SUCCESS
```
[the remark this answers is not in the document]
Comment (xbauch) [reply testsuite/gnat2goto/tests/private_type_dec_array/test.out]: Ok, removed.

File: gnat2goto/driver/tree_walk.adb [hunk 6]
```diff
@@ -4252,7 +4252,12 @@ package body Tree_Walk is
    -----------------------
 
    function Do_Type_Reference (E : Entity_Id) return Irep is
-      (Make_Symbol_Type (Identifier => Unique_Name (E)));
+   begin
+      if not Global_Symbol_Table.Contains (Intern (Unique_Name (E))) then
+         Declare_Itype (E);
```
[the remark this answers is not in the document]
Comment (xbauch) [reply gnat2goto/driver/tree_walk.adb]: Ok, removed.

File: gnat2goto/driver/arrays.adb [hunk 7]
```diff
@@ -944,12 +944,10 @@ package body Arrays is
       return List_Element (Struct_Component, Last_Cursor);
    end Get_Last_Index_Component;
 
-   function Get_Data_Component (Array_Struct : Irep;
-                                A_Symbol_Table : Symbol_Table)
-                                return Irep
+   function Get_Data_Component (Array_Struct : Irep) return Irep
```
[the remark this answers is not in the document]
Comment (xbauch) [reply gnat2goto/driver/arrays.adb]: Un-removed.

File: gnat2goto/driver/arrays.adb [hunk 8]
```diff
@@ -994,4 +992,20 @@ package body Arrays is
...
 
+   function Get_Data_Member (Array_Struct : Irep)
+                             return Irep
+   is
+      Data_Member : constant Irep :=
+        Get_Data_Component (Array_Struct);
+   begin
+      return Make_Member_Expr (Compound         => Array_Struct,
+                               Source_Location  =>
+                                 Get_Source_Location (Array_Struct),
```
[the remark this answers is not in the document]
Comment (xbauch) [reply gnat2goto/driver/arrays.adb]: Removed.

File: gnat2goto/driver/arrays.adb [hunk 9]
```diff
@@ -918,24 +914,21 @@ package body Arrays is
...
           Do_Expression (First (Expressions (N)))));
 
-   function Get_First_Index_Component (Array_Struct : Irep;
-                                       A_Symbol_Table : Symbol_Table)
+   function Get_First_Index_Component (Array_Struct : Irep)
                                        return Irep
    is
       Array_Struct_Type : constant Irep :=
-        Follow_Symbol_Type (Get_Type (Array_Struct), A_Symbol_Table);
+        Follow_Symbol_Type (Get_Type (Array_Struct), Global_Symbol_Table);
```
[the remark this answers is not in the document]
Comment (xbauch) [reply gnat2goto/driver/arrays.adb]: Added.

File: gnat2goto/driver/arrays.adb [hunk 10]
```diff
@@ -958,33 +951,29 @@ package body Arrays is
       return List_Element (Struct_Component, Last_Cursor);
    end Get_Data_Component;
 
-   function Get_First_Index (Array_Struct : Irep; Source_Loc : Source_Ptr;
-                             A_Symbol_Table : Symbol_Table)
-                             return Irep
+   function Get_First_Index (Array_Struct : Irep) return Irep
```
[the remark this answers is not in the document]
Comment (xbauch) [reply gnat2goto/driver/arrays.adb]: Removed.

File: gnat2goto/driver/arrays.ads (hunk 2 above)
Comment (xbauch) [reply gnat2goto/driver/arrays.ads]: Changed.

File: gnat2goto/driver/arrays.adb (hunk 3 above)
Comment (xbauch) [reply gnat2goto/driver/arrays.adb]: Removed.

File: gnat2goto/driver/arrays.adb (hunk 4 above)
Comment (xbauch) [reply gnat2goto/driver/arrays.adb]: Removed.

File: gnat2goto/driver/tree_walk.adb [hunk 11]
```diff
@@ -679,6 +679,15 @@ package body Tree_Walk is
    is
    begin
       if Ekind (Etype (Name (N))) in Array_Kind then
+         declare
+            Lhs_Type : constant Entity_Id := Etype (Name (N));
+         begin
```
Comment (hannes-steffenhagen-diffblue) [inline gnat2goto/driver/tree_walk.adb]: Could you leave a comment here explaining what's going on here for future reference?

Comment (xbauch) [reply gnat2goto/driver/tree_walk.adb]: Done.
